ci: macOS verification is failing due to std::filesystem requiring newer macOS

### What happened?

We need to target 10.15 or higher.

### Stack Trace

```
      x86_64-apple-darwin13.4.0-clang++ -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -fmessage-length=0 -isystem /private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/arrow-adbc-1.7.0.XXXXX.79CH5rsu/miniforge/envs/conda-source/include -D_FORTIFY_SOURCE=2 -isystem /private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/arrow-adbc-1.7.0.XXXXX.79CH5rsu/miniforge/envs/conda-source/include -I/private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/arrow-adbc-1.7.0.XXXXX.79CH5rsu/apache-arrow-adbc-1.7.0/python/adbc_driver_manager/adbc_driver_manager -I/private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/arrow-adbc-1.7.0.XXXXX.79CH5rsu/miniforge/envs/conda-source/include/python3.12 -c adbc_driver_manager/adbc_driver_manager.cc -o /var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/tmpylfmw3ja.build-temp/adbc_driver_manager/adbc_driver_manager.o -std=c++17
      adbc_driver_manager/adbc_driver_manager.cc:55:30: error: 'path' is unavailable: introduced in macOS 10.15 - see https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk
         55 | std::vector<std::filesystem::path> InternalAdbcParsePath(const std::string_view path);
            |                              ^
      /private/var/folders/vk/nx37ffx50hv5djclhltc26vw0000gn/T/arrow-adbc-1.7.0.XXXXX.79CH5rsu/miniforge/envs/conda-source/bin/../include/c++/v1/__filesystem/path.h:383:33: note: 'path' has been explicitly marked unavailable here
        383 | class _LIBCPP_EXPORTED_FROM_ABI path {
            |                                 ^
```

### How can we reproduce the bug?

_No response_

### Environment/Setup

_No response_

--- a/.github/workflows/nightly-verify.yml
+++ b/.github/workflows/nightly-verify.yml
@@ -165,6 +165,9 @@ jobs:
           # Rust uses a lot of disk space, free up some space
           # https://github.com/actions/runner-images/issues/2840
           sudo rm -rf "$AGENT_TOOLSDIRECTORY"
+          # Required for macOS
+          export CXXFLAGS="${CXXFLAGS} -D_LIBCPP_DISABLE_AVAILABILITY"
+          echo "CXXFLAGS=${CXXFLAGS}"
           ./arrow-adbc/dev/release/verify-release-candidate.sh $VERSION 0
 
       - name: Verify

--- a/.github/workflows/verify.yml
+++ b/.github/workflows/verify.yml
@@ -110,6 +110,9 @@ jobs:
           # Rust uses a lot of disk space, free up some space
           # https://github.com/actions/runner-images/issues/2840
           sudo rm -rf "$AGENT_TOOLSDIRECTORY"
+          # Required for macOS
+          export CXXFLAGS="${CXXFLAGS} -D_LIBCPP_DISABLE_AVAILABILITY"
+          echo "CXXFLAGS=${CXXFLAGS}"
           ./dev/release/verify-release-candidate.sh ${{ inputs.version }} ${{ inputs.rc }}
       - name: Verify
         if: matrix.os == 'windows-latest'

--- a/r/adbcdrivermanager/src/Makevars
+++ b/r/adbcdrivermanager/src/Makevars
@@ -16,7 +16,7 @@
 # under the License.
 
 CXX_STD = CXX17
-PKG_CPPFLAGS=-I../src/c/include -I../src/c -I../src/c/vendor -DADBC_EXPORT=""
+PKG_CPPFLAGS=-I../src/c/include -I../src/c -I../src/c/vendor -DADBC_EXPORT="" -D_LIBCPP_DISABLE_AVAILABILITY
 
 OBJECTS = driver_test.o \
     error.o \

